fix(linter): improve prefer_namespace_keyword rule

cblh opened the pull request:
fix: #4651
This commit enhances the `prefer_namespace_keyword` rule in the TypeScript linter: Add support for detecting and fixing nested module declarations (e.g., module A.B {})

- Introduce helper functions `is_nest_module`, `is_valid_module`, and `is_invalid_module` to improve code readability and maintainability
- Refactor the main `run` function to use these new helper functions
- Update test cases to cover nested module scenarios
- Improve error reporting for nested modules

cblh committed:
fix(linter): improve prefer_namespace_keyword rule

This commit enhances the `prefer_namespace_keyword` rule in the TypeScript linter:

- Introduce helper functions `is_nest_module`, `is_valid_module`, and `is_invalid_module` to improve code readability and maintainability
- Refactor the main `run` function to use these new helper functions
- Add support for detecting and fixing nested module declarations (e.g., `module A.B {}`)
- Update test cases to cover nested module scenarios
- Improve error reporting for nested modules

cblh committed:
test(linter): correct nested module handling in prefer_namespace_keyword rule

cblh committed:
chore(linter): Removed unnecessary context parameter from `is_valid_module` function.

cblh committed:
chore(linter): Refactor `prefer_namespace_keyword` rule in TypeScript linter

- Simplified `is_nest_module` function by removing redundant line breaks.

cblh committed:
fix(linter): remove redundant diagnostic return in `prefer_namespace_keyword` rule

- Removed an unnecessary return statement in the `prefer_namespace_keyword` rule to ensure proper execution flow and prevent premature diagnostic reporting.

rzvxa commented on crates/oxc_linter/src/rules/typescript/prefer_namespace_keyword.rs:
```diff
@@ -88,18 +105,32 @@ fn test() {
-			  declare module bar {}
+            module A {
+              module B {}
 			}
```
Your string formatting is still wrong, I guess you are using tabs over spaces - which is something I do as well - If you are using Vim write `:retab` to convert all of them to spaces.
On VSCode there is a built-in command called `Convert Indentations to Spaces` which you can use.

rzvxa commented on crates/oxc_linter/src/rules/typescript/prefer_namespace_keyword.rs:
```diff
@@ -34,13 +34,29 @@ declare_oxc_lint!(
-            || module.kind != TSModuleDeclarationKind::Module
-        {
+
+        if is_invalid_module(module) {
```
It would be super cool if we merge these 2 conditions together, However, if you like it this way feel free to ignore my comment. It is mostly a matter of taste.

cblh committed:
chore(linter): Refactor module validation functions in `prefer_namespace_keyword.rs`

- Renamed `is_invalid_module` to `is_valid_module` and updated its logic to check for valid modules.
- Renamed `is_valid_module` to `is_valid_module_node` to reflect its purpose more clearly.
- Modified `is_nest_module` to use the updated `is_valid_module_node`.
- Updated the `PreferNamespaceKeyword` rule to use the new validation logic, ensuring correct module validation and nesting checks.

rzvxa committed:
fmt

rzvxa reviewed (approved):
I've pushed a commit doing a small formatting unrelated to your changes(it was from an old test case).

Thanks for putting up with my nits❤️

rzvxa committed:
chore: update snapshot.

rzvxa committed:
fmt again

rzvxa committed:
Merge branch 'main' into fix/4651

cblh commented:
> I've pushed a commit doing a small formatting unrelated to your changes(it was from an old test case).我已经推送了一个提交，进行了与您的更改无关的小格式设置（它来自旧的测试用例）。
> 
> Thanks for putting up with my nits❤️谢谢你忍受我的烦恼❤️

Entirely understandable, and thank you for slogging through it with me!

cblh commented:
> I've pushed a commit doing a small formatting unrelated to your changes(it was from an old test case).
> 
> Thanks for putting up with my nits❤️

Entirely understandable, and thank you for slogging through it with me!
